generator/internal/discovery_to_proto.cc

```diff
@@ -60,6 +60,27 @@ google::cloud::StatusOr<std::string> GetPage(std::string const& url) {
   return rest::ReadAll(std::move(payload));
 }
 
+bool IsDiscoveryArrayType(nlohmann::json const& json) {
+  return json.contains("type") && json["type"] == "array" &&
+         json.contains("items");
+}
+
+bool IsDiscoveryMapType(nlohmann::json const& json) {
+  return json.contains("type") && json["type"] == "object" &&
+         json.contains("additionalProperties");
+}
+
+bool IsDiscoveryNestedType(nlohmann::json const& json) {
+  return json.contains("type") && json["type"] == "object" &&
+         json.contains("properties");
+}
+
+// std::set::merge isn't available until C++17.
```

Comment:
Ack.  I am not sure the single place where this function is called is more readable with this helper.  Your call.